### PR TITLE
Payment Mode in invoice detail added

### PR DESCRIPTION
**Pull request by ashdude1401**

### CODE REVIEW

File: src/main/java/com/darzee/shankh/service/OrderItemService.java [hunk 1]
```diff
@@ -88,17 +96,23 @@ public List<OrderItemDAO> createOrderItems(List<OrderItemDetailRequest> orderIte
...
-            List<PriceBreakupDAO> priceBreakupDAOList =
-                    priceBreakUpService.generatePriceBreakupList(itemDetail.getPriceBreakup(), orderItemDAO);
+            orderItemDAO = mapper
+                    .orderItemToOrderItemDAO(orderItemRepo.save(mapper.orderItemDAOToOrderItem(orderItemDAO,
+                            new CycleAvoidingMappingContext())), new CycleAvoidingMappingContext());
+
+            List<PriceBreakupDAO> priceBreakupDAOList = priceBreakUpService
+                    .generatePriceBreakupList(itemDetail.getPriceBreakup(), orderItemDAO);
+            priceBreakupDAOList = priceBreakUpService.savePriceBreakUp(priceBreakupDAOList);
+            orderItemDAO.setPriceBreakup(priceBreakupDAOList);
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderItemService.java:106]: why are we setting this price breakup in item?

File: src/main/java/com/darzee/shankh/service/OrderItemService.java [hunk 2]
```diff
@@ -113,13 +127,14 @@ public List<OrderItemDAO> createOrderItems(List<OrderItemDetailRequest> orderIte
...
+
             if (!CollectionUtils.isEmpty(itemDetail.getStitchOptionReferences())) {
                 stitchOptionService.addOrderItemId(itemDetail.getStitchOptionReferences(), orderItemDAO.getId());
             }
-            priceBreakupDAOList = priceBreakUpService.savePriceBreakUp(priceBreakupDAOList);
-            orderItemDAO.setPriceBreakup(priceBreakupDAOList);
+
             orderItemList.add(orderItemDAO);
         }
+
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderItemService.java:129]: remove extra lines

File: src/main/java/com/darzee/shankh/service/OrderItemService.java [hunk 3]
```diff
@@ -139,45 +181,60 @@ public OrderItemDAO updateOrderItem(Long orderItemId, OrderItemDetailRequest upd
...
             orderItem.setOrderItemStatus(status);
-//            ledgerService.handleBoutiqueLedgerOnOrderItemUpdation(orderItem.getOrder().getBoutique().getId(), initialState, status);
+            // ledgerService.handleBoutiqueLedgerOnOrderItemUpdation(orderItem.getOrder().getBoutique().getId(),
+            // initialState, status);
         }
+
         if (Boolean.TRUE.equals(updateItemDetail.getIsDeleted())) {
             orderItem.setIsDeleted(Boolean.TRUE);
-//            ledgerService.handleBoutiqueLedgerOnOrderItemDeletion(orderItem.getOrder().getBoutique().getId(), orderItem);
+            // ledgerService.handleBoutiqueLedgerOnOrderItemDeletion(orderItem.getOrder().getBoutique().getId(),
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderItemService.java:188]: remove the commented code.

File: src/main/java/com/darzee/shankh/service/OrderItemService.java [hunk 4]
```diff
@@ -227,18 +293,27 @@ public ResponseEntity<GetOrderItemResponse> getOrderItemDetails(Long boutiqueId,
...
         OutfitType outfitType = orderItem.getOutfitType();
         OutfitTypeService outfitTypeService = outfitTypeObjectService.getOutfitTypeObject(outfitType);
         String outfitImageLink = outfitTypeService.getOutfitImageLink();
+
         List<String> clothImagesReferenceIds = objectFilesService.getClothReferenceIds(orderItem.getId());
         List<FileDetail> clothImageFileDetails = getClothImageDetails(clothImagesReferenceIds);
+
         List<String> audioReferenceIds = objectFilesService.getAudioReferenceIds(orderItem.getId());
         List<FileDetail> audioFileDetails = getAudioDetails(audioReferenceIds);
+
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderItemService.java:279]: remove extra spacing

File: src/main/java/com/darzee/shankh/service/OrderItemService.java [hunk 5]
```diff
@@ -227,18 +293,27 @@ public ResponseEntity<GetOrderItemResponse> getOrderItemDetails(Long boutiqueId,
...
+        // Get the customer ID and get the overall measurement details for the customer
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderItemService.java:280]: remove these comments

File: src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java [hunk 6]
```diff
@@ -31,162 +31,266 @@
 @Service
 public class OrderOrderItemCommonService {
 
-    @Autowired
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java]: once done with changes, Select All and  press ctrl + alt + L in this file

File: src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java [hunk 7]
```diff
@@ -31,162 +31,266 @@
...
+                // get order details from request
+                OrderDetails orderDetails = orderCreationRequest.getOrderDetails();
+
+                // find or create new order
+                OrderDAO orderDAO = null;
+                try {
+                        orderDAO = orderService.findOrCreateNewOrder(orderDetails.getOrderId(),
+                                        orderDetails.getBoutiqueId(), orderDetails.getCustomerId());
+                } catch (NullPointerException e) {
+                        throw new RuntimeException("NullPointerException while finding or creating new order",
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java]: NullPointer is already a run time exception, why are we catching one null pointer exception and throwing another.

File: src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java [hunk 8]
```diff
@@ -31,162 +31,266 @@
...
-        if (orderDAO.getOrderAmount() == null) {
-            orderDAO.setOrderAmount(orderAmountDAO);
-            orderDAO = mapper.orderObjectToDao(orderRepo.save(mapper.orderaDaoToObject(orderDAO,
-                            new CycleAvoidingMappingContext())),
-                    new CycleAvoidingMappingContext());
+
+        @Transactional
+        public OrderSummary createOrder(OrderCreationRequest orderCreationRequest) {
+
+                // get order details from request
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java]: remove comments

File: src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java [hunk 9]
```diff
@@ -31,162 +31,266 @@
...
+
+                // create new order items
+                List<OrderItemDetailRequest> newItems = orderDetails.getOrderItems()
+                                .stream()
+                                .filter(itemDetail -> itemDetail.getId() == null)
+                                .collect(Collectors.toList());
+                List<OrderItemDAO> allOrderItems = null;
+                try {
+                        allOrderItems = orderItemService.createOrderItems(newItems, orderDAO);
+                } catch (Exception e) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java]: what exception is being catched here?

File: src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java [hunk 10]
```diff
@@ -31,162 +31,266 @@
...
+                        throw new RuntimeException("Exception while creating new order items",
+                                        e);
+                }
+                if (allOrderItems == null) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java]: is there a valid scenario where allOrderItems  could be null, I could see at max this could be an empty list, but not null, check accordingly.

File: src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java [hunk 11]
```diff
@@ -31,162 +31,266 @@
...
+                                .stream()
+                                .filter(itemDetail -> itemDetail.getId() != null)
+                                .collect(Collectors.toList());
+                List<OrderItemDAO> savedItems = new ArrayList<>();
+                for (OrderItemDetailRequest itemDetailRequest : alreadySavedItems) {
+                        OrderItemDAO orderItemDAO = null;
+                        try {
+                                orderItemDAO = orderItemService.updateOrderItem(itemDetailRequest.getId(),
+                                                itemDetailRequest);
+                        } catch (Exception e) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java]: which exeption is being catched here?

File: src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java [hunk 12]
```diff
@@ -31,162 +31,266 @@
...
+                                throw new RuntimeException("Exception while updating order item",
+                                                e);
+                        }
+                        if (orderItemDAO == null) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java]: can this be null ever?

File: src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java [hunk 13]
```diff
@@ -31,162 +31,266 @@
...
+                        if (!CollectionUtils.isEmpty(itemDetailRequest.getPriceBreakup())) {
+                                try {
+                                        // updating priceBreakup DAO
+                                        List<PriceBreakupDAO> updatedPriceBreakupDAOList = priceBreakUpService
+                                                        .updatePriceBreakups(itemDetailRequest.getPriceBreakup(),
+                                                                        orderItemDAO);
+
+                                        // updating priceBreakup DAO list of orderItemDAO
+                                        orderItemDAO.setPriceBreakup(updatedPriceBreakupDAOList);
+                                        ;
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderOrderItemCommonService.java]: remove extra ;, comments, lines, catch statements here.

File: src/main/java/com/darzee/shankh/service/OrderService.java [hunk 14]
```diff
@@ -110,8 +111,10 @@ public OrderDAO findOrCreateNewOrder(Long orderId, Long boutiqueId, Long custome
                 orderDAO = findOrder(orderId, boutiqueId);
                 return orderDAO;
             } catch (Exception e) {
+                System.out.println(e);
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderService.java:113]: remove

File: src/main/java/com/darzee/shankh/service/OrderService.java [hunk 15]
```diff
@@ -139,11 +141,7 @@ public OrderDAO createNewOrder(Long boutiqueId, Long customerId) {
         throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Invalid customer id or boutique id");
     }
 
-    public ResponseEntity<GetOrderResponse> getOrder(Long boutiqueId, String orderItemStatusList,
-                                                     String orderStatusList,
-                                                     Boolean priorityOrdersOnly, Long customerId,
-                                                     String deliveryDateFrom, String deliveryDateTill,
-                                                     Boolean paymentDue, Integer countPerPage, Integer pageCount) {
+    public ResponseEntity<GetOrderResponse> getOrder(Long boutiqueId, String orderItemStatusList,String orderStatusList, Boolean priorityOrdersOnly, Long customerId,String deliveryDateFrom, String deliveryDateTill, Boolean paymentDue, Integer countPerPage, Integer pageCount) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderService.java]: why have we done this?

File: src/main/java/com/darzee/shankh/service/OrderService.java [hunk 16]
```diff
@@ -376,6 +385,9 @@ public String generateInvoiceV2(OrderDAO orderDAO) {
         String customerName = customerDAO.constructName();
         BoutiqueDAO boutique = orderDAO.getBoutique();
         TailorDAO tailorDAO = boutique.getAdminTailor();
+
+        OrderAmountDAO orderAmountDAO = orderDAO.getOrderAmount();
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderService.java:344]: why is this added?

File: src/main/java/com/darzee/shankh/service/OrderService.java [hunk 17]
```diff
@@ -399,8 +411,19 @@ public ResponseEntity<InvoiceDetailResponse> getInvoiceDetail(Long orderId, Long
...
                     orderAmountDAO.getTotalAmount(), orderAmountDAO.getAmountRecieved(), orderDAO.getNonDeletedItems());
 
+            PaymentDAO paymentDAO = paymentService.getPaymentDAOByOrderId(
+                    orderDAO.getId());
+            Integer paymentMode;
+
+            if (paymentDAO != null) {
+                paymentMode = paymentDAO.getPaymentMode().getOrdinal();
+            } else {
+                paymentMode = null;
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderService.java:373]: no need of else, by default Integers are initialised as null

File: src/main/java/com/darzee/shankh/service/OrderService.java [hunk 18]
```diff
@@ -474,20 +497,20 @@ public ResponseEntity<InvoiceDetailResponse> getInvoiceDetail(Long orderId, Long
     // return orderAmount;
     // }
 
-//    private Double getAdvancePaid(List<PaymentDAO> payments) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderService.java:477]: remove the commented code

File: src/main/java/com/darzee/shankh/service/OrderService.java [hunk 19]
```diff
@@ -572,10 +600,13 @@ public OrderDAO updateOrderPostItemUpdation(Long orderId) {
         boolean allItemsDelivered = orderItems.stream()
                 .filter(item -> OrderItemStatus.DELIVERED.equals(item.getOrderItemStatus()))
                 .collect(Collectors.toList()).size() == orderItems.size();
+
         if (allItemsDelivered && !OrderStatus.DELIVERED.equals(orderDAO.getOrderStatus())) {
             orderDAO.setOrderStatus(OrderStatus.DELIVERED);
+
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/OrderService.java:539]: remove extra lines where ever applicable

File: src/main/java/com/darzee/shankh/service/PaymentService.java [hunk 20]
```diff
@@ -18,44 +22,58 @@ public class PaymentService {
...
+    public PaymentDAO recordPayment(Double amount, PaymentMode paymentMode, Boolean isAdvancePayment,OrderDAO order) {
         PaymentDAO paymentDAO = new PaymentDAO(amount, paymentMode, isAdvancePayment, order);
         paymentDAO = mapper.paymentToPaymentDAO(paymentRepo.save(mapper.paymentDAOToPayment(paymentDAO,
-                        new CycleAvoidingMappingContext())),
+                new CycleAvoidingMappingContext())),
                 new CycleAvoidingMappingContext());
         return paymentDAO;
     }
 
-//    public PaymentDAO updateAdvancePayment(OrderDAO order, Double finalAmount) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PaymentService.java:30]: remove commented code

File: src/main/java/com/darzee/shankh/service/PaymentService.java [hunk 21]
```diff
@@ -18,44 +22,58 @@ public class PaymentService {
...
             paymentRepo.save(mapper.paymentDAOToPayment(finalPayment, new CycleAvoidingMappingContext()));
             return finalPayment;
         }
         return null;
     }
+
+    public PaymentDAO getPaymentDAOByOrderId(Long orderId) {
+        Optional<Payment> payment = paymentRepo.findTopByOrderIdOrderByIdDesc(orderId);
+        if (payment.isPresent()) {
+            return mapper.paymentToPaymentDAO(payment.orElse(null), new CycleAvoidingMappingContext());
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PaymentService.java:71]: return mapper.paymentToPaymentDAO(payment.get(), new CycleAvoidingMappingContext());

File: src/main/java/com/darzee/shankh/service/PriceBreakUpService.java [hunk 22]
```diff
@@ -39,33 +39,60 @@ public List<PriceBreakupDAO> savePriceBreakUp(List<PriceBreakupDAO> priceBreakup
...
-                new CycleAvoidingMappingContext());
-        return priceBreakupDAOList;
 
+        try {
+            priceBreakupDAOList = mapper.priceBreakUpListToPriceBreakUpDAOList(
+                    priceBreakUpRepo.saveAll(mapper.priceBreakUpDAOListToPriceBreakUpList(priceBreakupDAOList,
+                            new CycleAvoidingMappingContext())),
+                    new CycleAvoidingMappingContext());
+
+        } catch (Exception e) {
```
Comment (raghav2499) [inline src/main/java/com/darzee/shankh/service/PriceBreakUpService.java:87]: what type of excepton are we catching here?

File: src/main/java/com/darzee/shankh/dao/OrderAmountDAO.java [hunk 23]
```diff
@@ -13,4 +13,5 @@ public class OrderAmountDAO {
     private Long id;
     private Double totalAmount = 0d;
     private Double amountRecieved = 0d;
+    private Integer paymentMode;
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/dao/OrderAmountDAO.java:16]: To provide payment inside paymentMode with OrderAmount . Payment mode is assoicated with OrderPayment detials .

File: src/main/java/com/darzee/shankh/dao/OrderItemDAO.java [hunk 24]
```diff
@@ -112,7 +112,7 @@ public boolean isQuantityUpdated(Integer value) {
     public boolean isMeasurementRevisionUpdated(Long value) {
         return (this.measurementRevision == null && value != null)
                 || (this.measurementRevision != null && value != null && !value.equals(
-                this.getMeasurementRevision().getId()));
+                    this.getMeasurementRevision().getId()));
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/dao/OrderItemDAO.java]: okk

File: .gitignore [hunk 25]
```diff
@@ -4,6 +4,8 @@ target/
 !**/src/main/**/target/
 !**/src/test/**/target/
 
+application.log
+
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply .gitignore:8]: okk .This line na   *.gz

File: src/main/java/com/darzee/shankh/controller/OrderController.java [hunk 26]
```diff
@@ -60,7 +62,8 @@ public ResponseEntity getOrderDetails(@PathVariable("id") Long boutiqueOrderId,
     public ResponseEntity recievePayment(@PathVariable("id") Long boutiqueOrderId,
                                          @RequestParam("boutique_id") Long boutiqueId,
                                          @Validated @RequestBody RecievePaymentRequest request) throws Exception {
-        return orderService.recieveOrderPayment(boutiqueOrderId, boutiqueId, request);
+        ResponseEntity response = orderService.recieveOrderPayment(boutiqueOrderId, boutiqueId, request);
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/controller/OrderController.java:51]: To improve readability and if future we can add try catch block in case of error .

File: src/main/java/com/darzee/shankh/service/OrderItemService.java [hunk 27]
```diff
@@ -186,6 +243,11 @@ public OrderItemDAO updateOrderItem(Long orderItemId, OrderItemDetailRequest upd
         }
         orderItem = mapper.orderItemToOrderItemDAO(orderItemRepo.save(mapper.orderItemDAOToOrderItem(orderItem,
                 new CycleAvoidingMappingContext())), new CycleAvoidingMappingContext());
+
+        if (orderItem.isPriceBreakupListUpdated(updateItemDetail.getPriceBreakup())) {
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/service/OrderItemService.java:232]: UpdatedOrderItem corresponds to the request which has the updated field .In updateOrderItem detail we are actually recursively checking each feild if it is different from the orderItem in our db with same Id .If it so, we are updating the Db orderItem .Priveously it was not checking the pricebreakup update .So , I added this code to check the update and change if required .

File: src/main/java/com/darzee/shankh/service/PaymentService.java (hunk 21 above)
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/service/PaymentService.java:71]: okk

File: src/main/java/com/darzee/shankh/service/PriceBreakUpService.java (hunk 22 above)
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/service/PriceBreakUpService.java:87]: It catching all type of exception which might be thrown from saveAll .But it need to be changed with DataAccessException e as it is interacting with db and most commons error can be of type DataAccessException while saving .

File: src/main/java/com/darzee/shankh/response/InvoiceDetailResponse.java [hunk 28]
```diff
@@ -41,6 +41,7 @@ public InvoiceDetailResponse( String invoiceDateTime, String boutiqueName,
         this.customerName = customerName;
         this.recieveDateTime = recieveDateTime;
         this.orderSummary = orderSummary;
+        this.paymentMode = Optional.ofNullable(paymentMode);
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/response/InvoiceDetailResponse.java:39]: If the paymentMode value pay or may not beb present .But I removed it as we are  we are manually sending null if it is not present

File: src/main/java/com/darzee/shankh/service/MeasurementService.java [hunk 29]
```diff
@@ -76,36 +76,57 @@ public class MeasurementService {
...
+            Long orderItemId,
+            Integer outfitTypeIndex,
+            String scale,
+            Boolean nonEmptyValuesOnly) throws Exception {
+        // Validate request parameters
         validateGetMeasurementRequestParams(customerId, measurementRevisionId, orderItemId, outfitTypeIndex, scale);
+
+        // Get Boutique ID
         Long boutiqueId = customerRepo.findById(customerId).get().getBoutique().getId();
+
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/service/MeasurementService.java:93]: Okk.Raghav .

File: src/main/java/com/darzee/shankh/service/MeasurementService.java [hunk 30]
```diff
@@ -271,25 +299,26 @@ private MeasurementRevisionsDAO addMeasurementRevision(MeasurementDetailsRequest
...
+            MeasurementScale scale,
+            Long boutiqueId, Boolean nonEmptyValuesOnly) throws Exception {
         OverallMeasurementDetails overallMeasurementDetails = new OverallMeasurementDetails();
         List<InnerMeasurementDetails> innerMeasurementDetailsList = new ArrayList<>();
         String measurementRevisionImgLink = null;
 
         if (CollectionUtils.isEmpty(revisionsDAO.getMeasurementValue())
                 && measurementRevisionService.measurementRevisionImageExists(revisionsDAO.getId())) {
-            measurementRevisionImgLink = measurementRevisionService.getMeasurementRevisionImageLink(revisionsDAO.getId());
-            innerMeasurementDetailsList = generateInnerMeasurementDetails(boutiqueId, revisionsDAO.getOutfitType(),
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/service/MeasurementService.java:291]: By mistake Ig .I have corrected it .

File: src/main/java/com/darzee/shankh/service/OrderItemService.java (hunk 2 above)
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/service/OrderItemService.java:129]: ok

File: src/main/java/com/darzee/shankh/service/OrderService.java (hunk 17 above)
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/service/OrderService.java:373]: okk

File: src/main/java/com/darzee/shankh/dao/OrderItemDAO.java [hunk 31]
```diff
@@ -154,4 +154,32 @@ public void setFormattedDeliveryDate() {
         DateTimeFormatter deliveryDateFormatter = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm a");
         this.formattedDeliveryDate = this.deliveryDate.format(deliveryDateFormatter);
     }
+
+
+    /*
+     * To check if price break up list is updated or not. If the list is empty or null, it will return true.
+     */
+    public boolean isPriceBreakupListUpdated(List<PriceBreakUpDetails> priceBreakupList) {
```
[the remark this answers is not in the document]
Comment (ashdude1401) [reply src/main/java/com/darzee/shankh/dao/OrderItemDAO.java:145]: Raghav check karne ke liye ki jo db orderItem hai and jo Order Item request mai hai usme priceBreakup same hai ki nahi .

**Review by raghav2499 — APPROVED** [review]
